src/test/java/com/example/gradecalculator/service/SubjectServiceTest.java

```diff
@@ -0,0 +1,162 @@
+package com.example.gradecalculator.service;
+
+import com.example.gradecalculator.entities.SchoolYear;
+import com.example.gradecalculator.entities.Subject;
+import com.example.gradecalculator.entities.User;
+import com.example.gradecalculator.entities.UserSubject;
+import com.example.gradecalculator.model.SubjectTO;
+import com.example.gradecalculator.repository.SchoolYearRepository;
+import com.example.gradecalculator.repository.SubjectRepository;
+import com.example.gradecalculator.repository.UserRepository;
+import com.example.gradecalculator.repository.UserSubjectRepository;
+import org.junit.jupiter.api.BeforeEach;
+import org.junit.jupiter.api.Test;
+import org.junit.jupiter.api.extension.ExtendWith;
+import org.mockito.InjectMocks;
+import org.mockito.Mock;
+import org.mockito.junit.jupiter.MockitoExtension;
+
+import java.time.LocalDate;
+import java.util.*;
+
+import static org.junit.jupiter.api.Assertions.*;
+import static org.mockito.Mockito.*;
+
+@ExtendWith(MockitoExtension.class)
+public class SubjectServiceTest {
```

Comment:
Tests cover mainly the positive outcome, also try to cover negativ cases (e.g. entry with id not found, delete not working)

Reply:
also what you can test:
- getAllSubjects: returned list should be alphabeticaly sorted
- saveSubjects: errors are added to list when already existing